lib/trino-memory-context/src/main/java/io/trino/memory/context/ThresholdLocalMemoryContext.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.trino.memory.context;
+
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+public class ThresholdLocalMemoryContext
+        implements LocalMemoryContext
+{
+    public static final long DEFAULT_SYNC_THRESHOLD = 65536;
```

Comment:
somewhere (query context? driver context? don't remember) there  is already a similar mechanism, with 1 MB threshold (AFAIR)

Reply:
There is some guaranteed memory somewhere - do you mean this one?

Reply:
Yeah - there is - but it is handled in the AggregatedMemoryContext in the synchronized section already [here](https://github.com/trinodb/trino/blob/98b9ccbefbd088f83fbfe93fbe72d6871ab500a8/lib/trino-memory-context/src/main/java/io/trino/memory/context/RootAggregatedMemoryContext.java#L33-L44). 
It is just to ensure we are reuturning non-blocked future. The accounting is still 100% correct - so it does not help with thread congestion.